pkg/cache/cache.go

```diff
@@ -0,0 +1,251 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Provides cache mechanism based on etags to lazily build, and/or cache
+// results from expensive operation such that those operations are not
+// repeated unnecessarily. The operations can be created as a tree, and
+// replaced dynamically as needed.
+//
+// # Dependencies and types of caches
+//
+// This package uses a source/transform/sink model of caches to build
+// the dependency tree, and can be used as follows:
+//   - [NewSource]: A source cache that recomputes the content every time.
+//   - [NewStaticSource]: A source cache that always produces the
+//     same content, it is only called once.
+//   - [NewTransformer]: A cache that transforms data from one format to
+//     another. It's only refreshed when the source changes.
+//   - [NewMerger]: A cache that aggregates multiple caches into one.
+//     It's only refreshed when the source changes.
+//   - [Replaceable]: A cache adapter that can ben atomically
+//     replaced with a new one, and saves the previous results in case an
+//     error pops-up.
+//
+// # Atomicity
+//
+// Most of the operations are not atomic/thread-safe, except for
+// [Replaceable.Replace] which can be performed while the objects
+// are being read.
+package cache
+
+import (
+	"fmt"
+	"sync/atomic"
+)
+
+// Result is the data returned from a call to a cache. It can either be
+// created with [NewResultData] if the call was a success, or
+// [NewResultError] if the call resulted in an error.
+type Result[T any] struct {
+	data T
+	etag string
+	err  error
+}
+
+// NewResultData creates a new [Result] for a successful operation.
+func NewResultData[T any](data T, etag string) Result[T] {
+	return Result[T]{
+		data: data,
+		etag: etag,
+	}
+}
+
+// NewResultError creates a new [Result] when an error has happened.
+func NewResultError[T any](err error) Result[T] {
+	return Result[T]{
+		err: err,
+	}
+}
+
+// Result can be treated as a [Interface] if necessary.
+func (r Result[T]) Data() Result[T] {
+	return r
+}
+
+// Split the result into its data, etag and error components.
+func (r Result[T]) Split() (T, string, error) {
+	return r.data, r.etag, r.err
+}
+
+// Interface is a cache that performs an action that will be cached. It
+// also returns an "etag" identifier to version the cache, so that the
+// caller can know if they have the most recent version of the cache
+// (and can decide to cache some operation based on that).
+//
+// The [NewMerger] and [NewTransformer] automatically handle
+// that for you by checking if the etag is updated before calling the
+// merging or transforming function.
+type Interface[T any] interface {
```

Comment:
I feel like we can come up with a less ambiguous name

Reply:
I can't find where right now, but rather than re-using the package name, you're supposed to use the name `Interface` when you would use the name of the package. So when used, it's not `cache.Cache`, but `cache.Interface` which is somewhat more useful.

Reply:
The concept implemented by this seems to be that of a handle? How about the name Handle?

Reply:
or Promise?

Reply:
Updated the name (and the name of the package). this becomes `cached.Data`.